### PR TITLE
[FIX] website_hr_recruitment: prevent applying for unpublished job

### PR DESCRIPTION
**Pull request by thle-odoo**
Issue:
------
When we have the link to the form to apply for a job, even if the job is not published, we can still apply.

Solution:
---------
Check that the job is published before the creation of the record.

opw-3331717

### CODE REVIEW

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 1]
```diff
@@ -29,6 +29,13 @@ class Applicant(models.Model):
 
     _inherit = 'hr.applicant'
 
+    @api.model
+    def create(self, vals):
+        job = self.env['hr.job'].browse(vals.get('job_id'))
+        if not job or not job.website_published:
+            return
```
Comment (rdeodoo) [inline addons/website_hr_recruitment/models/hr_recruitment.py:36]: at least raise an error for easier debugging and understanding of the error?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Do you know why isn't the model converter of the /apply controller is not going into a 403? That's the purpose of using modelConverter in route and IR.rules with `domain_force` set to `website_published=True`.
I guess you are preventing the creating through another way/route/rpc?

Anyway, the @odoo/rd-vidange team might want to review this one? I am not sure who is handling the `website_hr_recruitment`.

**Comment by thle-odoo** [comment]
The `/jobs/apply/` controller is called to load the application job form.
The `/website_form/model_name` controller is called when the form is submitted.
This flow inserts the new record in sudo (and therefore bypasses ir.rules).

I think that using the `website_form_input_filter` method can be useful by triggering a UserError so that the message is displayed on the website.

**Comment by kba-odoo** [comment]
robodoo r+

File: addons/website_hr_recruitment/models/hr_recruitment.py [hunk 2]
```diff
@@ -30,15 +31,17 @@ class Applicant(models.Model):
     _inherit = 'hr.applicant'
 
     def website_form_input_filter(self, request, values):
+        job = self.env['hr.job'].browse(values.get('job_id'))
+        if not job or not job.sudo().website_published:
+            raise UserError(_("You cannot apply for this job."))
```
Comment (rdeodoo) [inline addons/website_hr_recruitment/models/hr_recruitment.py:36]: Seems violent for stable, what about people with form to apply for a department only, without a job?

**Review by rdeodoo — CHANGES_REQUESTED** [review]
@robodoo r-

Just to be sure.. Are you removing the capability in a > 3 years old version to create job apply not linked to a job position but only to a department on purpose?

According to the commit title and desc, it's not what is trying to be achieved, you want to prevent unpublished job, but the code is actually not doing that, it's preventing unpublished job **and no job**.

Why not adding the raise condition inside the previous `if values.get('job_id')`?

If on purpose, you can r+ again, but doesn't seems very stable friendly, especially since you could fix only the problematic case?

Maybe in master you want to remove the jobless submission capability tho :shrug:

**Comment by thle-odoo** [comment]
@rdeodoo
The classic flow does not allow to apply to a department without a job.
However, it is indeed possible to realize this flow by using the "Apply for a Job" action with the editor (or some other way that we are not aware of).
Therefore, not having a job can happen.

Thanks for your remark !

**Comment by kba-odoo** [comment]
Thanks for your input indeed @rdeodoo :-) We didn't think at first that a manually added form could use that flow..

robodoo r+

**Comment by rdeodoo** [comment]
> Thanks for your input indeed @rdeodoo :-) We didn't think at first that a manually added form could use that flow..

No worries, all hail to the code owner ping :D

But on top of the manual form, there is all the custos / third party apps to consider in stable, especially in 14 :p 

Thanks for the change :+1:

**Pull request closed without merging**